lib/devcycle-ruby-server-sdk/eval_hooks_runner.rb

```diff
@@ -0,0 +1,135 @@
+      @original_error = original_error
+      @hook_context = hook_context
+    end
+
+    def to_s
+      msg = super
+      msg += "\nOriginal error: #{@original_error.message}" if @original_error
+      msg
+    end
+  end
+
+  class EvalHooksRunner
+    # @return [Array<EvalHook>] Array of eval hooks to run
+    attr_reader :eval_hooks
+
+    # Initializes the EvalHooksRunner with an optional array of eval hooks
+    # @param [Array<EvalHook>, nil] eval_hooks Array of eval hooks to run
+    def initialize(eval_hooks = [])
+      @eval_hooks = eval_hooks || []
+    end
+
+    # Runs all before hooks with the given context
+    # @param [HookContext] context The context to pass to the hooks
+    # @return [HookContext] The potentially modified context
+    # @raise [BeforeHookError] when a before hook fails
+    def run_before_hooks(context)
+      current_context = context
+      
+      @eval_hooks.each do |hook|
+        next unless hook.before
+
+        begin
+          result = hook.before.call(current_context)
+          # If the hook returns a new context, use it for subsequent hooks
+          current_context = result if result.is_a?(DevCycle::HookContext)
+        rescue => e
+          # Raise BeforeHookError to allow client to handle and skip after hooks
+          raise BeforeHookError.new(e.message, e, current_context)
+        end
+      end
+      
+      current_context
+    end
+
+    # Runs all after hooks with the given context
+    # @param [HookContext] context The context to pass to the hooks
+    # @return [void]
+    # @raise [AfterHookError] when an after hook fails
+    def run_after_hooks(context)
+      @eval_hooks.each do |hook|
+        next unless hook.after
+
+        begin
+          hook.after.call(context)
+        rescue => e
+          # Log error but continue with next hook
+          raise AfterHookError.new(e.message, e, context)
+        end
+      end
+    end
+
+    # Runs all finally hooks with the given context
+    # @param [HookContext] context The context to pass to the hooks
+    # @return [void]
+    def run_finally_hooks(context)
+      @eval_hooks.each do |hook|
+        next unless hook.on_finally
+
+        begin
+          hook.on_finally.call(context)
+        rescue => e
+          # Log error but don't re-raise to prevent blocking evaluation
+          warn "Error in finally hook: #{e.message}"
+        end
+      end
+    end
+
+    # Runs all error hooks with the given context and error
+    # @param [HookContext] context The context to pass to the hooks
+    # @param [Exception] error The error that occurred
+    # @return [void]
+    def run_error_hooks(context, error)
+      @eval_hooks.each do |hook|
+        next unless hook.error
+
+        begin
+          hook.error.call(context, error)
+        rescue => e
+          # Log error but don't re-raise to prevent blocking evaluation
+          warn "Error in error hook: #{e.message}"
+        end
+      end
+    end
```

Comment:
Do you mean these to run in the same thread or a new thread? This currently will run in the main thread of the app

Reply:
it also needs to be run before variable evaluation, as well as after variable evaluation

Reply:
If it works i guess we can fix it later and re-evaluate the threading issue if they appear